components/persistence/persistence-service-common/src/main/kotlin/net/corda/persistence/common/internal/EntitySandboxServiceImpl.kt

```
@@ -179,14 +179,38 @@ class EntitySandboxServiceImpl @Activate constructor(
             .toSet()
             .mapNotNull { getObserverFromClassName(it, ctx) }
             .groupBy { it.stateType }
-        ctx.putObjectByKey(SANDBOX_TOKEN_STATE_OBSERVERS, tokenStateObserverMap)
+
+        validateTokenStateObservers(tokenStateObserverMap)
+
+        // Create a map that contains a single UTXO token observer for each token state
+        // This away, people know that the there will never be more than one UTXO token observer
+        val validatedTokenStateObserverMap = tokenStateObserverMap.entries.associate { it.key to it.value.singleOrNull() }
+
+        ctx.putObjectByKey(SANDBOX_TOKEN_STATE_OBSERVERS, validatedTokenStateObserverMap)
         logger.debug {
             "Registered token observers: ${tokenStateObserverMap.mapValues { (_, observers) ->
                 observers.map { it::class.java.name }}
             }"
         }
     }
 
+    private fun validateTokenStateObservers(
+        tokenStateObserverMap: Map<Class<ContractState>, List<UtxoLedgerTokenStateObserver<ContractState>>>
+    ) {
+
+        tokenStateObserverMap.forEach { contractStateTypeToObservers ->
+            val numberOfObservers = contractStateTypeToObservers.value.size
+
+            if (numberOfObservers > 1) {
+                val observerTypes = contractStateTypeToObservers.value.map { it.stateType.toString() }
+                throw SandboxException(
```

Not sure if we should throw a sandbox exception or an illegal state. Can you check how we do it for registering vault named query factories, since we should try keep them consistent. I'll look later but I'm out atm.

---

If you are referring to the file `VaultNamedQueryRegistryImpl.kt`, it uses `IllegalArgumentException`.

---

I did not manage to find the file you are talking about. But illegal state makes more sense.

---

👍